operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/ResourceIDMatcherDiscriminator.java

```diff
@@ -5,21 +5,36 @@
 
 import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.javaoperatorsdk.operator.processing.event.ResourceID;
+import io.javaoperatorsdk.operator.processing.event.source.Cache;
 
 public class ResourceIDMatcherDiscriminator<R extends HasMetadata, P extends HasMetadata>
     implements ResourceDiscriminator<R, P> {
 
+
+  private final String eventSourceName;
   private final Function<P, ResourceID> mapper;
 
   public ResourceIDMatcherDiscriminator(Function<P, ResourceID> mapper) {
+    this(null, mapper);
+  }
+
+  public ResourceIDMatcherDiscriminator(String eventSourceName, Function<P, ResourceID> mapper) {
+    this.eventSourceName = eventSourceName;
     this.mapper = mapper;
   }
 
+  @SuppressWarnings("unchecked")
   @Override
   public Optional<R> distinguish(Class<R> resource, P primary, Context<P> context) {
     var resourceID = mapper.apply(primary);
-    return context.getSecondaryResourcesAsStream(resource)
-        .filter(resourceID::isSameResource)
-        .findFirst();
+    if (eventSourceName != null) {
```

Comment:
In the keycloak case there is only a single Service event source, so eventSourceName doesn't seem required in all cases.  Could it test for, or catch the exception, if there are multiple sources for the given type before resorting to the list?

Reply:
makes sense, changed, pls check if this looks to you